### ISSUE TITLE
twisted.internet.posixbase._Waker not defined on Jython

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3718]
|<img alt="pjenvey's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| pjenvey reported|
|-|-|
|Trac ID|trac#3718|
|Type|defect|
|Created|2009-03-30 19:30:31Z|
|Branch|https://github.com/twisted/twisted/tree/jython-waker-3718|

posixbase defines a _Waker depending on the value of twisted.python.runtime.platformType, which is 'java' on Jython. There's no _Waker definition for 'java'

```
$ jython -W all ./bin/trial --reporter=bwverbose --unclean-warnings twisted
<snip>
Traceback (most recent call last):
  File "./bin/trial", line 24, in <module>
    run()
  File "/Users/pjenvey/src/java/jython/jython-twisted/src/Twisted/twisted/scripts/trial.py", line 361, in run
    test_result = trialRunner.run(suite)
  File "/Users/pjenvey/src/java/jython/jython-twisted/src/Twisted/twisted/trial/runner.py", line 813, in run
    return self._runWithoutDecoration(test)
  File "/Users/pjenvey/src/java/jython/jython-twisted/src/Twisted/twisted/trial/runner.py", line 839, in _runWithoutDecoration
    run()
  File "/Users/pjenvey/src/java/jython/jython-twisted/src/Twisted/twisted/trial/runner.py", line 834, in <lambda>
    run = lambda: suite.run(result)
  File "/Users/pjenvey/src/java/jython/jython-twisted/src/Twisted/twisted/trial/runner.py", line 293, in run
    self._bail()
  File "/Users/pjenvey/src/java/jython/jython-twisted/src/Twisted/twisted/trial/runner.py", line 275, in _bail
    from twisted.internet import reactor
  File "/Users/pjenvey/src/java/jython/jython-twisted/src/Twisted/twisted/internet/reactor.py", line 38, in <module>
    selectreactor.install()
  File "/Users/pjenvey/src/java/jython/jython-twisted/src/Twisted/twisted/internet/selectreactor.py", line 200, in install
    reactor = SelectReactor()
  File "/Users/pjenvey/src/java/jython/jython-twisted/src/Twisted/twisted/internet/selectreactor.py", line 72, in __init__
    posixbase.PosixReactorBase.__init__(self)
  File "/Users/pjenvey/src/java/jython/jython-twisted/src/Twisted/twisted/internet/posixbase.py", line 168, in __init__
    ReactorBase.__init__(self)
  File "/Users/pjenvey/src/java/jython/jython-twisted/src/Twisted/twisted/internet/base.py", line 467, in __init__
    self._initThreads()
  File "/Users/pjenvey/src/java/jython/jython-twisted/src/Twisted/twisted/internet/base.py", line 887, in _initThreads
    self.installWaker()
  File "/Users/pjenvey/src/java/jython/jython-twisted/src/Twisted/twisted/internet/posixbase.py", line 204, in installWaker
    self.waker = _Waker(self)
NameError: global name '_Waker' is not defined
```

Attachments:

* [waker.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f62/f62f9738bb8f669c90a6f1a0ab28c7e3ac3f5c6b/d7b88aca7760690b8f71fdcffc4502d263f0b78e.patch) (1088 bytes) - added by sanxiyn on 2009-06-17 14:18:50Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3718 3718
type__defect defect
reporter__pjenvey pjenvey
priority__normal normal
milestone__ 
branch__branches_jython_waker_3718 branches/jython-waker-3718
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1238441431000000 1238441431000000
changetime__1266765646000000 1266765646000000
version__None None
owner__ 
cc__thijs cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_2"></a>|@thijstriemstra commented|
|-|-|

Updating ticket description, fixing typo (Waker instead of Walker)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_3"></a>|@thijstriemstra commented|
|-|-|

The buildslave is [reporting](http://buildbot.twistedmatrix.com/builders/ubuntu64-jython2.5-select/builds/147/steps/trial/logs/stdio) the same error message now.

**Comment by twisted-trac** [comment]
|<img alt="sanxiyn's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|sanxiyn commented|
|-|-|

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="sanxiyn's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|sanxiyn commented|
|-|-|

Oops, os.pipe is present on win32. Is "I am a workaround for the lack of pipes on win32" accurate?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve set owner to sanxiyn|
|-|-|

Yeah, pipes are available on windows, but are not selectable (iirc). So you need to check for something else, unfortunately. And indeed the docstring of `Win32Waker` should be updated.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [28218]) Branching to 'jython-waker-3718'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [28219]) Use the newly renamed _SocketWaker on Windows, _PipeWaker everywhere else

refs [#3718](https://github.com/twisted/twisted/issues/3718)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun removed owner|
|-|-|

Fixed some of that stuff.  Jython can [run the tests](http://buildbot.twistedmatrix.com/builders/ubuntu64-jython2.5-select/builds/84) in the branch.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_11"></a>|@thijstriemstra commented|
|-|-|

I{self-pipe trick<http://cr.yp.to/docs/selfpipe.html>} should probably use U instead of I.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_12"></a>|@thijstriemstra commented|
|-|-|

Replying to [exarkun](#note_10):
> Fixed some of that stuff.  Jython can [run the tests](http://buildbot.twistedmatrix.com/builders/ubuntu64-jython2.5-select/builds/84) in the branch.
>

Trial is dying with:

```
Exception in thread "MainThread" java.lang.OutOfMemoryError: Java heap space
```

Exarkun, could you increase the memory? From #jython:

```
pjenvey: default is 512mb, you can do jython -J-Xmx1024m
pjenvey: JAVA_MEM=-Xmx1024m jython should work too
```

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_13"></a>|@thijstriemstra set owner to @exarkun|
|-|-|

I see you added a 1024, but looks like it's still [not enough](http://buildbot.twistedmatrix.com/builders/ubuntu64-jython2.5-select/builds/85/steps/trial/logs/stdio). Can you double it?

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_14"></a>|@thijstriemstra commented|
|-|-|

I just ran it (on that buildslave) with 2048 and it still fails. I guess the buildslave, or jython support, is not an option until we take a look at trial's memory usage.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

One possible explanation for the memory usage is that all the failures trial is keeping around are responsible.  This might be helped by throwing away the actual failures and just keeping the string that will be reported at the end.

If this is the cause, then another possible work-around would be to run a limited subset of the tests on Jython until those all pass, then expand that set and repeat.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun removed owner|
|-|-|

I set up the builder to just run `twisted.test` for now.  Let's see [what happen](http://buildbot.twistedmatrix.com/builders/ubuntu64-jython2.5-select/builds/94/steps/trial/logs/stdio)s...

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_17"></a>|@therve set owner to @exarkun|
|-|-|

Looks better, I guess. +1.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun set status to `closed`|
|-|-|

(In [28292]) Merge jython-waker-3718

Author: sanxiyn, exarkun
Reviewer: therve
Fixes: [#3718](https://github.com/twisted/twisted/issues/3718)

Change the waker selection code to be aware of non-Windows, non-POSIX platforms (such as
Java/Jython).

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
